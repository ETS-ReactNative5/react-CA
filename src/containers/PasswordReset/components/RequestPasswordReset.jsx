import React, { Component } from 'react'
import PropTypes from 'prop-types'

import Aux from 'components/Hoc/Aux/Aux'
import Button from 'components/Core/Button/Button'
import FormGroup from 'components/Core/FormGroup/FormGroup'
import { Row, Col } from 'components/Core/Section/Section'
import { validators } from 'utils/formatters/validators'

import mailIcon from 'assets/images/homepage/mail.svg'
import arrowRight from 'assets/images/global/arrow-right.svg'
import styles from '../PasswordReset.scss'


class RequestPasswordReset extends Component {
  state = {
    emailAddress: null,
    formIsValid: false,
  }

  handleEmailAddressChange = emailAddress => this.setState(ps => ({ ...ps, emailAddress }))
  handleIsValid = formIsValid => this.setState(ps => ({ ...ps, formIsValid }))

  inputs = [{
    id: 'email',
    placeholder: 'Email',
    value: this.state.emailAddress,
    icon: mailIcon,
    handler: this.handleEmailAddressChange,
    validators: validators.globals.email,
  }]

  renderPasswordResetRequestForm = () => {
    const { loading, onSubmit } = this.props
    const isDisabled = loading || !this.state.formIsValid

    return (
      <Aux>
        <Row><h1 className={styles.header}>Password Reset</h1></Row>
        <Row>
          <Col sm='12' md='8' lg='8'>
            <p>Enter the email address linked to your account!</p>
            <FormGroup
              className={styles.formGroup}
              inputs={this.inputs}
              isValid={this.handleIsValid}
            />
          </Col>
        </Row>
        <Row>
          <Button
            onClick={() => onSubmit(this.state.emailAddress)}
            disabled={isDisabled}
            icon={arrowRight}
            text='Reset My Password'
            primary
          />
        </Row>
      </Aux>
    )
  }

  renderSuccessMessage = () => (
    <Aux>
      <Row><h1 className={styles.header}>Reset password email sent!</h1></Row>
      <Row>
        <p className={styles.header}>
          To reset your password, you must click the link sent to your email address.
        </p>
      </Row>
    </Aux>
  )

  render() {
    const { emailSent } = this.props
    if (emailSent) return this.renderSuccessMessage()
    return this.renderPasswordResetRequestForm()
  }
}

RequestPasswordReset.propTypes = {
  onSubmit: PropTypes.func.isRequired,
  emailSent: PropTypes.bool.isRequired,
  loading: PropTypes.bool.isRequired,
}


export default RequestPasswordReset
