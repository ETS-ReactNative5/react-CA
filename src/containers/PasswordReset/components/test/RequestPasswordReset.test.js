import React from 'react'
import { shallow } from 'enzyme'

import RequestPasswordReset from '../RequestPasswordReset'

const componentProps = {
  onSubmit: jest.fn(),
  emailSent: false,
  loading: false,
}

describe('<RequestPasswordReset />', () => {
  it('renders correctly', () => {
    const wrapper = shallow(<RequestPasswordReset {...componentProps} />)
    expect(wrapper).toMatchSnapshot()
  })

  it('does not allow submission when loading prop is true', () => {
    const testProps = { ...componentProps, loading: true }
    const wrapper = shallow(<RequestPasswordReset {...testProps} />)
    const button = wrapper.find('Button')
    expect(button).toBeDisabled()
  })

  it('does not allow submission when invalid email is supplied', () => {
    const wrapper = shallow(<RequestPasswordReset {...componentProps} />)
    const button = wrapper.find('Button')

    wrapper.setState({ formIsValid: false })
    wrapper.update()

    expect(button).toBeDisabled()
  })

  it('sends a valid email address to the onSubmit callback', () => {
    const wrapper = shallow(<RequestPasswordReset {...componentProps} />)
    const button = wrapper.find('Button')

    wrapper.setState({ formIsValid: true, emailAddress: 'bacon@vegan.com' })
    wrapper.update()
    button.simulate('click')

    expect(componentProps.onSubmit).toHaveBeenCalledWith('bacon@vegan.com')
  })

  it('displays a success page when emailSent prop is true', () => {
    const testProps = { ...componentProps, emailSent: true }
    const wrapper = shallow(<RequestPasswordReset {...testProps} />)
    const text = 'To reset your password, you must click the link sent to your email address.'

    expect(wrapper.find('p').text()).toEqual(text)
  })
})
