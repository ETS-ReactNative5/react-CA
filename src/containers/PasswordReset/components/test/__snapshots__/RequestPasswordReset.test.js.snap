// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RequestPasswordReset /> renders correctly 1`] = `
<Aux>
  <Row
    className=""
  >
    <h1
      className="header"
    >
      Password Reset
    </h1>
  </Row>
  <Row
    className=""
  >
    <Col
      className=""
      lg="8"
      lg-offset={null}
      md="8"
      md-offset={null}
      sm="12"
      sm-offset={null}
      xl={null}
      xl-offset={null}
      xs={null}
      xs-offset={null}
    >
      <p>
        Enter the email address linked to your account!
      </p>
      <FormGroup
        className="formGroup"
        inputs={
          Array [
            Object {
              "handler": [Function],
              "icon": "mail.svg",
              "id": "email",
              "placeholder": "Email",
              "validators": Array [
                Object {
                  "error": Object {
                    "level": 3,
                    "message": "Please enter a valid email address!",
                  },
                  "regex": /\\^\\[a-zA-Z0-9\\.\\\\-_\\]\\+@\\[a-zA-Z0-9\\\\-_\\]\\+\\\\\\.\\[a-zA-Z0-9\\.\\\\-_\\]\\+\\$/,
                },
              ],
              "value": null,
            },
          ]
        }
        isValid={[Function]}
      />
    </Col>
  </Row>
  <Row
    className=""
  >
    <Button
      className=""
      disabled={true}
      icon="arrow-right.svg"
      iconOnly={false}
      middle={true}
      nocaps={false}
      onClick={[Function]}
      outline={false}
      primary={true}
      secondary={false}
      small={false}
      text="Reset My Password"
      textOnly={false}
      tiny={false}
    />
  </Row>
</Aux>
`;
